### PR TITLE
Persist the parent in ParentBuildAction

### PR DESCRIPTION
**Pull request by aheubusch**
Allows to retrieve the parent build in case Jenkins has restarted.

### CODE REVIEW

**Comment by olivergondza** [comment]
I presume the motivation is to make the matrix build able to survive the restart, right?. In such case, testcase is definitely needed (`RestartableJenkinsRule`).

File: src/main/java/hudson/matrix/MatrixConfiguration.java [hunk 1]
```diff
@@ -492,17 +491,36 @@ public boolean scheduleBuild(List<? extends Action> actions, Cause c) {
...
                     ? (MatrixBuild)currentExecutor.getCurrentExecutable() : null;
+            parentId = parent != null ? parent.getExternalizableId() : null;
         }
         
         public boolean shouldSchedule(List<Action> actions) {
             return true;
         }
+        
+        public MatrixBuild getMatrixBuild()
+        {
```
Comment (olivergondza) [inline src/main/java/hudson/matrix/MatrixConfiguration.java]: Please, respect surrounding code style (`() {`).

**Comment by aheubusch** [comment]
Fixed code style and added a test case that tests the intended behavior.

File: src/main/java/hudson/matrix/MatrixConfiguration.java [hunk 2]
```diff
@@ -492,17 +491,34 @@ public boolean scheduleBuild(List<? extends Action> actions, Cause c) {
...
             this.parent = currentExecutor != null 
                     ? (MatrixBuild)currentExecutor.getCurrentExecutable() : null;
+            parentId = parent != null ? parent.getExternalizableId() : null;
         }
         
         public boolean shouldSchedule(List<Action> actions) {
             return true;
         }
+        
+        public MatrixBuild getMatrixBuild() {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/matrix/MatrixConfiguration.java]: would be nice to have some Javadoc and annotations

File: src/main/java/hudson/matrix/MatrixConfiguration.java [hunk 3]
```diff
@@ -492,17 +491,34 @@ public boolean scheduleBuild(List<? extends Action> actions, Cause c) {
...
+            if (parent == null && parentId != null) {
+                try {
+                    parent = (MatrixBuild)Run.fromExternalizableId(parentId);
```
Comment (oleg-nenashev) [inline src/main/java/hudson/matrix/MatrixConfiguration.java]: Unchecked class cast

File: src/main/java/hudson/matrix/MatrixConfiguration.java (hunk 2 above)
Comment (aheubusch) [inline src/main/java/hudson/matrix/MatrixConfiguration.java]: @oleg-nenashev I am not experienced with with java and jenkins. What annotations would make sense?

**Comment by olivergondza** [comment]
I am working on polishing this. Got destructed for days porting this over to plugin pom v2 and getting tests to run again.

**Pull request merged**